src/NumericalAlgorithms/RootFinding/TOMS748.hpp

```diff
@@ -6,15 +6,465 @@
+  // to the root.
+  //
+  // Note: `bracket` will only modify slots marked as `true` in
+  //       `incomplete_mask`
+  const T tol_batch = std::numeric_limits<T>::epsilon() * static_cast<T>(2);
+
+  // If the interval [a,b] is very small, or if c is too close
+  // to one end of the interval then we need to adjust the
+  // location of c accordingly. This is:
+  //
+  //   if ((b - a) < 2 * tol * a) {
+  //     c = a + (b - a) / 2;
+  //   } else if (c <= a + fabs(a) * tol) {
+  //     c = a + fabs(a) * tol;
+  //   } else if (c >= b - fabs(b) * tol) {
+  //     c = b - fabs(b) * tol;
+  //   }
+  const T a_filt = simd::fma(fabs(a), tol_batch, a);
+  const T b_filt = simd::fnma(fabs(b), tol_batch, b);
+  const T b_minus_a = b - a;
+  c = simd::select(
+      (static_cast<T>(2) * tol_batch * a > b_minus_a) and incomplete_mask,
+      simd::fma(b_minus_a, static_cast<T>(0.5), a),
+      simd::select(c <= a_filt, a_filt, simd::select(c >= b_filt, b_filt, c)));
+
+  // Invoke f(c):
+  T fc = f(c);
+
+  // if we have a zero then we have an exact solution to the root:
+  const auto fc_is_zero_mask = (fc == static_cast<T>(0));
+  if (const auto mask = fc_is_zero_mask and incomplete_mask;
+      UNLIKELY(simd::any(mask))) {
+    a = simd::select(mask, c, a);
+    fa = simd::select(mask, static_cast<T>(0), fa);
+    d = simd::select(mask, static_cast<T>(0), d);
+    fd = simd::select(mask, static_cast<T>(0), fd);
+    if (UNLIKELY(simd::all(mask or not incomplete_mask))) {
+      return;
+    }
+  }
+
+  // Non-zero fc, update the interval:
+  //
+  // Note: unlike Boost, we assume fa*fc doesn't overflow. This speeds up the
+  // code quite a bit.
+  //
+  // Boost code is:
+  // if (boost::math::sign(fa) * boost::math::sign(fc) < 0) {...} else {...}
+  using simd::sign;
+  const auto sign_mask = AssumeFinite
+                             ? (fa * fc < static_cast<T>(0))
+                             : (sign(fa) * sign(fc) < static_cast<T>(0));
+  const auto mask_if =
+      (sign_mask and (not fc_is_zero_mask)) and incomplete_mask;
+  d = simd::select(mask_if, b, d);
+  fd = simd::select(mask_if, fb, fd);
+  b = simd::select(mask_if, c, b);
+  fb = simd::select(mask_if, fc, fb);
+
+  const auto mask_else =
+      ((not sign_mask) and (not fc_is_zero_mask)) and incomplete_mask;
+  d = simd::select(mask_else, a, d);
+  fd = simd::select(mask_else, fa, fd);
+  a = simd::select(mask_else, c, a);
+  fa = simd::select(mask_else, fc, fa);
+}
+
+template <bool AssumeFinite, class F, class T, class Tol>
+std::pair<T, T> toms748_solve(F f, const T& ax, const T& bx, const T& fax,
+                              const T& fbx, Tol tol,
+                              const simd::mask_type_t<T>& ignore_filter,
+                              size_t& max_iter) {
+  // Main entry point and logic for Toms Algorithm 748
+  // root finder.
+  if (UNLIKELY(simd::any(ax >= bx))) {
+    throw std::domain_error("Lower bound is larger than upper bound");
```

Comment:
I'm assuming something subtle breaks in the implementation. The Boost implementation doesn't allow it so I just did that rather than possibly creating some weird edge case